src/main/java/com/faforever/client/update/ClientUpdateServiceImpl.java

```diff
@@ -30,67 +30,91 @@
     this.platformService = platformService;
     this.applicationContext = applicationContext;
+    this.preferencesService = preferencesService;
+
+    currentVersion = defaultString(Version.getCurrentVersion(), DEVELOPMENT_VERSION_STRING);
+    log.info("Current version: {}", currentVersion);
 
-    currentVersion = new ComparableVersion(
-        defaultString(Version.VERSION, DEVELOPMENT_VERSION_STRING)
-    );
-    logger.info("Current version: {}", currentVersion);
+    CheckForUpdateTask task = applicationContext.getBean(CheckForUpdateTask.class);
+    this.updateInfoFuture = taskService.submitTask(task).getFuture();
   }
 
   /**
-   * Returns information about an available update. Returns {@code null} if no update is available.
+   * Returns information about an available mandatory update. Returns {@code null} if no update is available.
    */
-  public void checkForUpdateInBackground() {
-    CheckForUpdateTask task = applicationContext.getBean(CheckForUpdateTask.class);
-    task.setCurrentVersion(currentVersion);
+  @Override
+  public CompletableFuture<UpdateInfo> checkForMandatoryUpdate() {
+    return updateInfoFuture;
+  }
 
-    taskService.submitTask(task).getFuture().thenAccept(updateInfo -> {
+  /**
+   * Creates an update notification with actions to download and install latest release
+   */
+  public void checkForRegularUpdateInBackground() {
+    updateInfoFuture.thenAccept(updateInfo -> {
       if (updateInfo == null) {
         return;
       }
-      notificationService.addNotification(new PersistentNotification(
-          i18n.get("clientUpdateAvailable.notification", updateInfo.getName(), formatSize(updateInfo.getSize(), i18n.getUserSpecificLocale())),
-          INFO, asList(
-          new Action(i18n.get("clientUpdateAvailable.downloadAndInstall"), event -> downloadAndInstallInBackground(updateInfo)),
-          new Action(i18n.get("clientUpdateAvailable.releaseNotes"), Action.Type.OK_STAY,
-              event -> platformService.showDocument(updateInfo.getReleaseNotesUrl().toExternalForm())
-          )))
-      );
+
+      try {
+        // no async call because this task runs asynchronously already
+        ReleaseInfo latestRelease = preferencesService.getRemotePreferences().getLatestRelease();
+
+        if (!Version.shouldUpdate(getCurrentVersion(), latestRelease.getVersion())) {
+          return;
+        }
+
+        notificationService.addNotification(new PersistentNotification(
+            i18n.get("clientUpdateAvailable.notification", updateInfo.getName(), formatSize(updateInfo.getSize(), i18n.getUserSpecificLocale())),
+            INFO, asList(
+            new Action(i18n.get("clientUpdateAvailable.downloadAndInstall"), event -> downloadAndInstallInBackground(updateInfo)),
+            new Action(i18n.get("clientUpdateAvailable.releaseNotes"), Action.Type.OK_STAY,
+                event -> platformService.showDocument(updateInfo.getReleaseNotesUrl().toExternalForm())
+            )))
```

Comment:
is it well formatted?

Reply:
Yes, I think so. Technically I did not touch this code, but it got intended, that's why it's showing up.